pkg/controller/replicaset/replica_set.go

```diff
@@ -198,6 +203,17 @@ func (rsc *ReplicaSetController) Run(workers int, stopCh <-chan struct{}) {
 // getPodReplicaSet returns the replica set managing the given pod.
 // TODO: Surface that we are ignoring multiple replica sets for a single pod.
 func (rsc *ReplicaSetController) getPodReplicaSet(pod *api.Pod) *extensions.ReplicaSet {
+	// look up in the cache, if cached and the cache is valid, just return cached value
+	// Note: if we create a new RS that is also matching a given pod, it will just be ignored
+	// (assuming that the previously matching RS didn't changed), but we sort overlapping RSs
+	// by creation timestamp, so it's no problem. If we change the sort order, we may need
+	// change the code here too.
```

Comment:
You are right. :+1: So, I think we should invalid all the cache once a RS's selector updated. It could be done in  `UpdateFunc: `https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L124

A selector update operation could make all the cache invaild. But I don't think it's a big problem since it doesn't happen too often. WDTK?

In addition, theoretically this PR could only improve 3-5 times since ususally, a RS will have a replica number 3-5, and a  selector update operation could make all the cache invaild. I think we may eventually need to add namespace index.